bluetooth: Adapt for A2DP Source

jialu522 opened the pull request:
bug: v/74718

jialu522 committed:
bluetooth: Adapt for A2DP Source

bug: v/74718

Signed-off-by: jialu <jialu@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -44,6 +44,7 @@
 #include "a2dp_codec.h"
 
 #define A2DP_PEER_ENDPOINT_MAX 10
+#define A2DP_BUF_SIZE 672
```
Use a buffer size that is large enough for the worst case.
And add a comment to illustrate how it is calculated.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -45,6 +45,9 @@
 
 #define A2DP_PEER_ENDPOINT_MAX 10
 
+/* A2DP_TX_BUF_SIZE = frame_len * MAX_PCM_FRAME_NUM_PER_TICK + 1(nb_frames) + AVDTP_hander(12) + L2CAP_hander(4) */
```
For SBC, A2DP_TX_BUF_SIZE should not be less than xxxxxx.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -45,6 +45,9 @@
 
 #define A2DP_PEER_ENDPOINT_MAX 10
 
+/* for SBC: A2DP_TX_BUF_SIZE >= frame_len * MAX_PCM_FRAME_NUM_PER_TICK + 1(nb_frames) + AVDTP_hander(12) + L2CAP_hander(4) */
```
nb_frames(1)

jialu522 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -45,6 +45,9 @@
 
 #define A2DP_PEER_ENDPOINT_MAX 10
 
+/* for SBC: A2DP_TX_BUF_SIZE >= frame_len * MAX_PCM_FRAME_NUM_PER_TICK + 1(nb_frames) + AVDTP_hander(12) + L2CAP_hander(4) */
```
typo: header

jialu522 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -1784,6 +1786,10 @@ bt_status_t bt_sal_a2dp_source_send_data(bt_controller_id_t id, bt_address_t* re
     }
 
+    if (nbytes > A2DP_TX_BUF_SIZE - 12 - 4) {
+        BT_LOGW("%s, Message too long", __func__);
```
Print `nbytes` and `A2DP_TX_BUF_SIZE - 12 - 4` as well
e.g.,
`"%s, nbytes(%d) too long, max: %d"`

gzh-terry replied:
`LOGD`

And change to verbose in the future.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -1784,6 +1786,10 @@ bt_status_t bt_sal_a2dp_source_send_data(bt_controller_id_t id, bt_address_t* re
     }
 
+    if (nbytes > A2DP_TX_BUF_SIZE - 12 - 4) {
+        BT_LOGW("%s, Message too long, nbytes = %d, A2DP_TX_BUF_SIZE - 12 - 4 = %d", __func__, nbytes, A2DP_TX_BUF_SIZE - 12 - 4);
```
`BT_LOGD("%s, message too long, nbytes: %d, max: %d", __func__, nbytes, (A2DP_TX_BUF_SIZE - 12 - 4));`

jialu522 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -1784,6 +1786,10 @@ bt_status_t bt_sal_a2dp_source_send_data(bt_controller_id_t id, bt_address_t* re
     }
 
+    if (nbytes > A2DP_TX_BUF_SIZE - 12 - 4) {
+        BT_LOGW("%s, Message too long, nbytes = %d, A2DP_TX_BUF_SIZE - 12 - 4 = %d", __func__, nbytes, A2DP_TX_BUF_SIZE - 12 - 4);
```
`if (nbytes > (A2DP_TX_BUF_SIZE - 12 - 4)) {`

jialu522 replied:
done

chengkai15 commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -111,8 +114,7 @@ static bool flag_is_conn_none(struct zblue_a2dp_info_t* a2dp_info)
 
-NET_BUF_POOL_DEFINE(bt_a2dp_tx_pool, CONFIG_BT_MAX_CONN,
-    BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU),
+NET_BUF_POOL_DEFINE(bt_a2dp_tx_pool, CONFIG_BT_MAX_CONN, A2DP_TX_BUF_SIZE,
```
the fact that  a A2DP  frame size  is less than a L2cap buffer size,  it seems that we do not set it as A2DP_TX_BUF_SIZE

jialu522 replied:
A2DP with a Bitpool value of 32 requires a receive buffer of 600+ bytes. Our configuration (CONFIG_BT_L2CAP_TX_MTU=498) is insufficient for A2DP. In reality, the packet length checked by L2CAP is L2CAP_BR_DEFAULT_MTU=672. Additionally, the A2DP packets are too large, causing the L2CAP layer to return errors directly.

chengkai15 commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -123,7 +125,7 @@ static struct bt_a2dp_codec_ie sbc_src_ie = {
         0xFF, /* Block length: 4/8/12/16, subbands:4/8, Allocation Method: SNR, Londness */
         0x02, /* min bitpool */
-        0x35, /* max bitpool */
+        0x20, /* max bitpool */
```
why do we neeed to limit smalller bitpool?

jialu522 replied:
When Bitpool=53, the A2DP packet length exceeds 672 bytes, causing audio packet transmission failures.

zhangyuan376 commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -143,10 +145,10 @@ static struct bt_a2dp_codec_ie src_sbc_ie_default[] = {
             0x15, /* Block length: 4/8/12/16, subbands:4/8, Allocation Method: SNR, Londness */
             0x02, /* min bitpool */
-            0x35, /* max bitpool */
+            0x20, /* max bitpool */
```
Why limit the bitpool to 20?

jialu522 replied:
When Bitpool=53, the A2DP packet length exceeds 672 bytes, causing audio packet transmission failures.

jialu522 committed:
bluetooth: Optimize A2DP list

bug: v/74718

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: Change the resource release logic for the A2DP list

bug: v/74718

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: Fix configuration setting errors.

bug: v/74718

When configuring, the local codec, remote codec, and selected codec must be passed. Using the preferred codec as the selected codec may cause errors because the bitpool value is a subset of the local codec, remote codec, and preferred codec.

Signed-off-by: jialu <jialu@xiaomi.com>

jialu522 committed:
bluetooth: Free the list storing the peer device's codec information before the set configuration process.

bug: v/74718

Before the set configuration process, compare the local codec, remote codec, and preferred codec. After the comparison, delete the locally stored codec information to reduce memory usage.

Signed-off-by: jialu <jialu@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -60,9 +60,9 @@ struct zblue_a2dp_info_t {
-    uint8_t peer_endpoint_count;
-    struct bt_a2dp_codec_ie peer_sbc_capabilities[A2DP_PEER_ENDPOINT_MAX];
-    struct bt_a2dp_ep found_peer_endpoint[A2DP_PEER_ENDPOINT_MAX];
+    struct bt_a2dp_ep* found_peer_endpoint;
```
`selected_peer_endpoint` sounds more reasonable.
If it refers to a final codec that is selected according to local, peer, and preferred codec.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -60,9 +60,9 @@ struct zblue_a2dp_info_t {
-    struct bt_a2dp_ep found_peer_endpoint[A2DP_PEER_ENDPOINT_MAX];
+    struct bt_a2dp_ep* found_peer_endpoint;
+    bt_list_t* peer_endpoint;
+    struct bt_a2dp_codec_cfg* config_endpoint;
```
the name shall be a `config` or `xxx_config` rather than a `config_endpoint`

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -599,8 +598,30 @@ static struct bt_sdp_attribute a2dp_sink_attrs[] = {
 
+static void a2dp_peer_endpoint_destory(void* data)
+{
+    struct bt_a2dp_ep* found_peer_endpoint = (struct bt_a2dp_ep*)data;
```
Not necessary to name as `found_peer_endpoint` since the function name is `a2dp_peer_endpoint_destory`

We can just name it `endpoint`

gzh-terry replied:
Alternatively, `peer_endpoint` to avoid freeing a static memory.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -599,8 +598,30 @@ static struct bt_sdp_attribute a2dp_sink_attrs[] = {
 static struct bt_sdp_record a2dp_sink_rec = BT_SDP_RECORD(a2dp_sink_attrs);
 #endif /* CONFIG_BLUETOOTH_A2DP_SINK */
 
+static void a2dp_peer_endpoint_destory(void* data)
```
Typo: `destroy`

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -599,8 +598,30 @@ static struct bt_sdp_attribute a2dp_sink_attrs[] = {
+    if (a2dp_info->found_peer_endpoint)
+        a2dp_peer_endpoint_destory(a2dp_info->found_peer_endpoint);
+
     free(data);
```
Do we need a
```
free(a2dp_info->config_endpoint->codec_config);
free(a2dp_info->config_endpoint)
```

jialu522 replied:
yes

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -715,52 +736,73 @@ static bt_status_t check_local_remote_codec_sbc(uint8_t* local_ie, uint8_t* remo
     bt_status_t flag;
+    bt_list_node_t* node;
+    bt_list_t* list;
+    struct bt_a2dp_ep* found_peer_endpoint;
```
`peer_endpoint` or `peer_ep` or `ep`
It only becomes `found_` after `success:`

jialu522 replied:
ack

gzh-terry replied:
Nothing changed bro.

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -715,52 +736,73 @@ static bt_status_t check_local_remote_codec_sbc(uint8_t* local_ie, uint8_t* remo
+    struct bt_a2dp_codec_cfg* preferred, struct bt_a2dp_codec_cfg* config)
 {
-    int index = 0;
     bt_status_t flag;
```
`flag` might be a `status` since the type is `bt_status_t`.

Typically `flag` is a bitwise value.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -1008,15 +1055,14 @@ static uint8_t bt_a2dp_discover_endpoint_cb(struct bt_a2dp* a2dp,
             return BT_A2DP_DISCOVER_EP_STOP;
 
+        struct bt_a2dp_ep* found_peer_endpoint = (struct bt_a2dp_ep*)malloc(sizeof(struct bt_a2dp_ep));
+        found_peer_endpoint->codec_cap = (struct bt_a2dp_codec_ie*)malloc(sizeof(struct bt_a2dp_codec_ie));
```
Any value for this `codec_cap`?

jialu522 replied:
*ep = found_peer_endpoint; the stack handles the assignment

gzh-terry replied:
Can we use a `zalloc` or `calloc`?
In case thae stack failed to assign values.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -996,8 +1045,6 @@ static struct bt_a2dp_stream_ops stream_ops = {
 static uint8_t bt_a2dp_discover_endpoint_cb(struct bt_a2dp* a2dp,
```
Break into small functions, each around 50 lines.

jialu522 replied:
done

jialu522 committed:
bluetooth: Optimize code

bug: v/74718

Too much duplicated code increases the number of code lines; refactor the duplicated code using macros.

Signed-off-by: jialu <jialu@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -78,6 +78,25 @@ struct zblue_a2dp_info_t {
 
 static bt_list_t* bt_a2dp_conn = NULL;
 
+#define BT_A2DP_SET_CONFIG(a2dp_info, local, preferred)                                                                    \
```
Try to use function rather than macro.

jialu522 replied:
ack

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -780,6 +846,13 @@ static void zblue_on_stream_configured(struct bt_a2dp_stream* stream)
     codec_config.codec_type = zephyr_codec_2_sal_codec(stream->local_ep->codec_type);
     codec_cfg = &stream->codec_config;
 
+    if (a2dp_info->config) {
```
Move these to a `a2dp_config_delete(a2dp_info)`? I saw it 3 times.

jialu522 replied:
This is not an issue; we have implemented the minimum required changes.

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
Move this into `a2dp_info_create`, right before a2dp_info_destroy
And use zalloc or calloc to avoid massive initializations.

jialu522 replied:
This is not an issue; we have implemented the minimum required changes.

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```
if (!a2dp_info) {
    a2dp_info_create(a2dp);
    return; /* wait the peer device to start a sep discovery */
}

BT_LOGD("a2dp_info already exists");
flag_set(a2dp_info, A2DP_STATE_BIT_SIG_CONN);
if (a2dp_info->int_acp == A2DP_INT) {
    a2dp_info->peer_endpoint = bt_list_new(a2dp_peer_endpoint_destroy);
    bt_a2dp_discover(a2dp, &bt_discover_param);
}

```

jialu522 replied:
This is not an issue; we have implemented the minimum required changes.

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
Add error detections for `err`

jialu522 replied:
No failure scenarios exist

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
Looks very similar with those in `zblue_on_connected`?

jialu522 replied:
This is not an issue; we have implemented the minimum required changes.

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -993,12 +1047,37 @@ static struct bt_a2dp_stream_ops stream_ops = {
 #endif
 };
 
+static bt_status_t bt_A2DP_set_config(struct zblue_a2dp_info_t* a2dp_info, struct bt_a2dp_ep* local,
```
`A2DP` -> `a2dp`

jialu522 replied:
done

gzh-terry commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -993,12 +1047,37 @@ static struct bt_a2dp_stream_ops stream_ops = {
+
+    while (local_index < preferred_count) {
+        a2dp_info->config = (struct bt_a2dp_codec_cfg*)malloc(sizeof(struct bt_a2dp_codec_cfg));
+        a2dp_info->config->codec_config = (struct bt_a2dp_codec_ie*)malloc(sizeof(struct bt_a2dp_codec_ie));
```
Would be better if we have some method to do this work without many `malloc` and `free`

jialu522 replied:
ack

gzh-terry reviewed (approved):
LGTM now

chengkai15 commented on Kconfig:
```diff
@@ -296,13 +296,27 @@ config BLUETOOTH_A2DP_SOURCE
 
+config ZBLUE_A2DP_SOURCE_POOL_SIZE
+	int "A2DP source pool size"
+	default 3280
```
why default value is 3280? 

could we calulate  from ZBLUE_A2DP_SBC_MAX_BIT_POOL?

jialu522 replied:
This is incorrect - it should be buffer size, len=672 (L2CAP MTU) - 12 (AVDTP header)

chengkai15 commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -143,10 +142,10 @@ static struct bt_a2dp_codec_ie src_sbc_ie_default[] = {
         .len = 4,
         .codec_ie = {
-            0x28, /* 16000 | 32000 | 44100 | 48000 | mono | dual channel | stereo | join stereo */
+            0x21, /* 16000 | 32000 | 44100 | 48000 | mono | dual channel | stereo | join stereo */
```
why change it from 0x28 to 0x21?

jialu522 replied:
Joint Stereo provides better sound quality

chengkai15 commented on service/stacks/zephyr/sal_a2dp_interface.c:
```diff
@@ -598,8 +596,27 @@ static struct bt_sdp_attribute a2dp_sink_attrs[] = {
+        return;
+
+    if (peer_endpoint->codec_cap)
+        free(peer_endpoint->codec_cap);
```
could we need to unset peer_endpoint->codec_cap to avoid some one reference it?

jialu522 replied:
No need to set to NULL, as this function is called during bt_list_remove/free operations

hyson710 commented on Kconfig:
```diff
@@ -296,13 +296,27 @@ config BLUETOOTH_A2DP_SOURCE
+	default 32
+	depends on BLUETOOTH_STACK_BREDR_ZBLUE
+	help
+		A2DP sbc codec max bit pool 1~53
```
@jialu522 need to verify the max bit pool after L2CAP is ready

jialu522 replied:
ack
